src/main/java/com/atlassian/db/replica/internal/aurora/AuroraClusterDiscovery.java

```diff
@@ -12,12 +14,22 @@
 
 public final class AuroraClusterDiscovery {
     private final AuroraConnectionDetails auroraConnectionDetails;
+    private final SuppliedCache<Collection<Database>> discoveredReplicasCache;
 
-    public AuroraClusterDiscovery(AuroraConnectionDetails auroraConnectionDetails) {
+    private AuroraClusterDiscovery(
+        AuroraConnectionDetails auroraConnectionDetails,
+        SuppliedCache<Collection<Database>> discoveredReplicasCache
+    ) {
         this.auroraConnectionDetails = auroraConnectionDetails;
+        this.discoveredReplicasCache = discoveredReplicasCache;
     }
 
     public Collection<Database> getReplicas(Supplier<Connection> connectionSupplier) {
+        return discoveredReplicasCache.get(() -> fetchReplicas(connectionSupplier))
```

Comment:
fetchReplicas if it's null then fetch it again?

Reply:
yeah, when there will be no value in the cache'a then we need to fetch them. We cannot return empty collection here because it will be used in consistency checking -> empty replicas collection means that we are consistent with no replicas... which is incorrect.

Reply:
Yes, but why do we call fetchReplicas(connectionSupplier) twice and expect different results?

```discoveredReplicasCache.get(() -> fetchReplicas(connection supplier))``` loads the cache from the supplier if the cache is empty. If the cache load fails, then we call the supplier again and return the results without caching it.

Reply:
If cache loads data from supplier when is empty, then why `.get(supplier)` returns Optional? I guess it may only happen in case of some error. WDYT should i create custom exception here?

Reply:
I don't know why it uses optional.

Reply:
Ok, I see good solution. I will return empty collection and delegate to the client (consistency checking) decision. I guess for empty collection `isConsistent` should return false (we are not consistent with replicas where there are no replicas).

Reply:
If we have 0 replicas then we are always consistent, no?

Reply:
I also think that an empty collection of replicas should indicate that we are consistent - not the other way around

Reply:
IMO we shouldn't call `isConsistent` when there are no replicas. When we check replicas consistency and there are no replicas for me it sounds like error. At the end it doesn't matter because we forward move to main when replicas are unavailable/inconsistent.

Reply:
> When we check replicas consistency and there are no replicas for me it sounds like error.
Why error? If all replicas fail then there will be some time without any replica (especially for the single replica setups).

empty collection = no replicas
exception = can't check consistency (by default DualConnection switches to the main database)